File: iOS/Hitze-V/Hitze-V/Services/DashboardDataService.swift
import Foundation
import CoreLocation

final class DashboardDataService {
    private let urlSession: URLSession
    private let defaultGeoSphereBaseURL = "https://warnungen.zamg.at/wsapp/api/getWarningsForCoords"
    private let openMeteoBaseURL = "https://api.open-meteo.com/v1/forecast"
    private let viennaTimeZone = TimeZone(identifier: "Europe/Vienna") ?? .current

    init(urlSession: URLSession = .shared) {
        self.urlSession = urlSession
    }

    private var geoSphereBaseURL: String {
        if AppFeatureFlags.enableCustomGeoSphereURLSetting {
            let configuredURL = UserDefaults.standard.string(forKey: "network.customGeoSphereUrl")?
                .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            if !configuredURL.isEmpty {
                return configuredURL
            }
        }

        return defaultGeoSphereBaseURL
    }

    private var customGeoSphereURL: String? {
        guard AppFeatureFlags.enableCustomGeoSphereURLSetting else {
            return nil
        }

        let configuredURL = UserDefaults.standard.string(forKey: "network.customGeoSphereUrl")?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return configuredURL.isEmpty ? nil : configuredURL
    }

    func fetchSnapshot(for coordinate: CLLocationCoordinate2D) async throws -> WorksiteSnapshot {
        async let geoSphere = fetchGeoSphere(for: coordinate)
        async let meteo = fetchOpenMeteo(for: coordinate)

        let geoResult = try await geoSphere
        let meteoResult = try await meteo

        let now = Date()
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = viennaTimeZone
        var forecasts: [DailyForecast] = []
        
        let targetDates = (0..<4).compactMap { calendar.date(byAdding: .day, value: $0, to: now) }
        
        for (i, targetDate) in targetDates.enumerated() {
            var uvMax: Double?
            var tempMax: Double?
            
            if i < meteoResult.daily.time.count {
                uvMax = meteoResult.daily.uvIndexMax[i]
                tempMax = meteoResult.daily.apparentTemperatureMax[i]
            }
            
            let severity = highestHazardSeverity(for: targetDate, warnings: geoResult.warnings)
            let warningTimeRanges = warningTimeRanges(for: targetDate, warnings: geoResult.warnings)
            
            forecasts.append(DailyForecast(
                date: targetDate,
                severity: severity,
                apparentTemperatureMax: tempMax,
                uvIndexMax: uvMax,
                warningTimeRanges: warningTimeRanges
            ))
        }

        let todayForecast = forecasts.first

        return WorksiteSnapshot(
            municipalityID: geoResult.municipalityID,
            municipalityName: geoResult.municipalityName,
            severity: todayForecast?.severity ?? .none,
            uvIndex: todayForecast?.uvIndexMax ?? nil,
            apparentTemperature: todayForecast?.apparentTemperatureMax ?? nil,
            forecasts: forecasts,
            updatedAt: Date()
        )
    }

    private func fetchGeoSphere(for coordinate: CLLocationCoordinate2D) async throws -> GeoSphereResolvedState {
        let url: URL
        if let customURL = customGeoSphereURL {
            guard let parsedURL = URL(string: customURL) else {
                throw DashboardDataError.invalidRequest
            }
            url = parsedURL
        } else {
            var components = URLComponents(string: geoSphereBaseURL)
            components?.queryItems = [
                URLQueryItem(name: "lat", value: String(format: "%.6f", coordinate.latitude)),
                URLQueryItem(name: "lon", value: String(format: "%.6f", coordinate.longitude))
            ]

            guard let parsedURL = components?.url else {
                throw DashboardDataError.invalidRequest
            }
            url = parsedURL
        }

        NSLog("GeoSphere request URL: %@", url.absoluteString)
        let (data, response) = try await performGET(url)
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            if let httpResponse = response as? HTTPURLResponse {
                let bodyPreview = String(data: data.prefix(600), encoding: .utf8) ?? "<non-utf8 body>"
                NSLog("GeoSphere HTTP failure: %ld for %@", httpResponse.statusCode, url.absoluteString)
                NSLog("GeoSphere HTTP failure body preview: %@", bodyPreview)
            }
            if let municipalityNotFoundMessage = municipalityNotFoundMessage(from: data) {
                throw DashboardDataError.municipalityNotFound(message: municipalityNotFoundMessage)
            }
            throw DashboardDataError.network(message: "GeoSphere HTTP Fehler")
        }

        let decoded: GeoSphereLookupResponse
        do {
            decoded = try JSONDecoder().decode(GeoSphereLookupResponse.self, from: data)
        } catch {
            debugGeoSphereDecodingFailure(error: error, data: data, url: url, response: httpResponse)
            throw DashboardDataError.invalidGeoSphereResponse
        }

        if decoded.type.lowercased() == "error" {
            throw DashboardDataError.municipalityNotFound(
                message: decoded.msg ?? "Could not find municipal for coords."
            )
        }

        guard let municipality = decoded.properties?.location?.properties,
              let municipalityID = municipality.gemeindenr.value,
              !municipalityID.isEmpty else {
            throw DashboardDataError.invalidGeoSphereResponse
        }

        let municipalityName = municipality.name ?? "Gemeinde \(municipalityID)"
        let warnings = decoded.properties?.warnings ?? []

        if warnings.isEmpty {
            NSLog("GeoSphere response for municipality %@ contains 0 warnings", municipalityID)
        } else {
            NSLog("GeoSphere response for municipality %@ contains %ld warnings", municipalityID, warnings.count)
        }

        return GeoSphereResolvedState(
            municipalityID: municipalityID,
            municipalityName: municipalityName,
            warnings: warnings
        )
    }

    private func municipalityNotFoundMessage(from data: Data) -> String? {
        guard let decoded = try? JSONDecoder().decode(GeoSphereLookupResponse.self, from: data) else {
            return nil
        }

        guard decoded.type.lowercased() == "error" else {
            return nil
        }

        let message = decoded.msg?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !message.isEmpty else {
            return "Could not find municipal for coords."
        }

        let normalizedMessage = message.lowercased()
        if normalizedMessage.contains("could not find municipal for coords") ||
            normalizedMessage.contains("could not find municipality") ||
            normalizedMessage.contains("municipal for coords") {
            return message
        }

        return nil
    }
    
    private func parseGeosphereTimestamp(_ value: String?) -> Date? {
        guard let value = value?.trimmingCharacters(in: .whitespacesAndNewlines), !value.isEmpty else { return nil }
        
        if let timeInterval = TimeInterval(value) {
            // ZamG timestamps are often in milliseconds
            if timeInterval > 100000000000 {
                return Date(timeIntervalSince1970: timeInterval / 1000.0)
            }
            return Date(timeIntervalSince1970: timeInterval)
        }
        
        let isoFormatter = ISO8601DateFormatter()
        if let date = isoFormatter.date(from: value) {
            return date
        }

        let localFormatter = DateFormatter()
        localFormatter.locale = Locale(identifier: "de_AT_POSIX")
        localFormatter.timeZone = viennaTimeZone
        localFormatter.dateFormat = "dd.MM.yyyy HH:mm"
        if let date = localFormatter.date(from: value) {
            return date
        }
        
        return nil
    }

    private func highestHazardSeverity(for targetDate: Date, warnings: [GeoSphereWarning]) -> HazardSeverity {
        let highestHeat = heatWarnings(for: targetDate, warnings: warnings)
            .map(\.level)
            .max() ?? 0

        if highestHeat > 0 {
            return HazardSeverity.heat(from: highestHeat)
        }
        
        return .none
    }

    private func warningTimeRanges(for targetDate: Date, warnings: [GeoSphereWarning]) -> [WarningTimeRange] {
        let calendar = gregorianViennaCalendar()
        let startOfDay = calendar.startOfDay(for: targetDate)
        guard let endOfDay = calendar.date(byAdding: DateComponents(day: 1, second: -1), to: startOfDay) else {
            return []
        }

        return heatWarnings(for: targetDate, warnings: warnings)
            .map { warning in
                WarningTimeRange(
                    start: max(warning.start ?? startOfDay, startOfDay),
                    end: min(warning.end ?? endOfDay, endOfDay)
                )
            }
            .sorted { $0.start < $1.start }
            .reduce(into: [WarningTimeRange]()) { merged, range in
                guard let lastRange = merged.last else {
                    merged.append(range)
                    return
                }

                if range.start > lastRange.end {
                    merged.append(range)
                } else {
                    merged[merged.count - 1] = WarningTimeRange(
                        start: lastRange.start,
                        end: max(lastRange.end, range.end)
                    )
                }
            }
    }

    private func heatWarnings(for targetDate: Date, warnings: [GeoSphereWarning]) -> [DailyHeatWarning] {
        let calendar = gregorianViennaCalendar()
        let startOfTargetDay = calendar.startOfDay(for: targetDate)
        guard let endOfTargetDay = calendar.date(byAdding: DateComponents(day: 1, second: -1), to: startOfTargetDay) else {
            return []
        }

        return warnings.compactMap { warning in
            let properties = warning.properties
            let startDate = parseGeosphereTimestamp(resolvedWarningStart(properties))
            let endDate = parseGeosphereTimestamp(resolvedWarningEnd(properties))

            let isActiveOnTargetDate: Bool
            switch (startDate, endDate) {
            case let (startDate?, endDate?):
                isActiveOnTargetDate = !(endDate < startOfTargetDay || startDate > endOfTargetDay)
            case let (startDate?, nil):
                isActiveOnTargetDate = startDate <= endOfTargetDay
            case let (nil, endDate?):
                isActiveOnTargetDate = endDate >= startOfTargetDay
            case (nil, nil):
                isActiveOnTargetDate = true
            }

            guard isActiveOnTargetDate, isHeatWarning(properties) else {
                return nil
            }

            let level = resolvedWarningLevel(properties)
            return DailyHeatWarning(level: level, start: startDate, end: endDate)
        }
    }

    private func resolvedWarningStart(_ properties: GeoSphereWarningProperties) -> String? {
        let candidates = [
            properties.start,
            properties.begin,
            properties.rawinfo?.start,
            properties.rawfinfo?.start
        ]

        return candidates.compactMap { $0?.trimmingCharacters(in: .whitespacesAndNewlines) }.first { !$0.isEmpty }
    }

    private func resolvedWarningEnd(_ properties: GeoSphereWarningProperties) -> String? {
        let candidates = [
            properties.end,
            properties.rawinfo?.end,
            properties.rawfinfo?.end
        ]

        return candidates.compactMap { $0?.trimmingCharacters(in: .whitespacesAndNewlines) }.first { !$0.isEmpty }
    }

    private func resolvedWarningLevel(_ properties: GeoSphereWarningProperties) -> Int {
        let candidates = [
            properties.warnstufeid,
            properties.wlevel,
            properties.rawinfo?.wlevel,
            properties.rawfinfo?.wlevel
        ]
        .compactMap { $0 }

        return candidates.max() ?? 0
    }

    private func isHeatWarning(_ properties: GeoSphereWarningProperties) -> Bool {
        let typeCandidates = [
            properties.wtype,
            properties.warntypid,
            properties.rawinfo?.wtype,
            properties.rawfinfo?.wtype
        ]
        .compactMap { $0 }

        if !typeCandidates.isEmpty {
            return typeCandidates.contains(6)
        }

        guard let textualType = properties.warningType?.lowercased() else {
            return false
        }

        return textualType.contains("hitze") || textualType.contains("heat")
    }

    private func gregorianViennaCalendar() -> Calendar {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = viennaTimeZone
        return calendar
    }

    private func fetchOpenMeteo(for coordinate: CLLocationCoordinate2D) async throws -> OpenMeteoForecastResponse {
        var components = URLComponents(string: openMeteoBaseURL)
        components?.queryItems = [
            URLQueryItem(name: "latitude", value: String(format: "%.6f", coordinate.latitude)),
            URLQueryItem(name: "longitude", value: String(format: "%.6f", coordinate.longitude)),
            URLQueryItem(name: "daily", value: "uv_index_max,apparent_temperature_max"),
            URLQueryItem(name: "timezone", value: "Europe/Vienna"),
            URLQueryItem(name: "forecast_days", value: "4")
        ]

        guard let url = components?.url else {
            throw DashboardDataError.invalidRequest
        }

        let (data, response) = try await performGET(url)
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw DashboardDataError.network(message: "Open-Meteo HTTP Fehler")
        }

        let decoded = try JSONDecoder().decode(OpenMeteoForecastResponse.self, from: data)
        return decoded
    }

    private func performGET(_ url: URL) async throws -> (Data, URLResponse) {
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 10
        request.cachePolicy = .reloadIgnoringLocalCacheData

        do {
            return try await urlSession.data(for: request)
        } catch let error as URLError where error.code == .cancelled {
            throw CancellationError()
        } catch is CancellationError {
            throw CancellationError()
        } catch {
            NSLog("Network transport error for %@: %@", url.absoluteString, error.localizedDescription)
            throw DashboardDataError.network(message: error.localizedDescription)
        }
    }

    private func debugGeoSphereDecodingFailure(
        error: Error,
        data: Data,
        url: URL,
        response: HTTPURLResponse
    ) {
        NSLog("GeoSphere decode failed for URL: %@", url.absoluteString)
        NSLog("GeoSphere HTTP status: %ld", response.statusCode)
        NSLog("GeoSphere decoding error: %@", String(describing: error))

        if let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            NSLog("GeoSphere top-level keys: %@", object.keys.sorted().joined(separator: ", "))

            if let firstKey = object.keys.sorted().first,
               let nestedObject = object[firstKey] as? [String: Any] {
                NSLog(
                    "GeoSphere nested keys for %@: %@",
                    firstKey,
                    nestedObject.keys.sorted().joined(separator: ", ")
                )
            }
        }

        let bodyPreview = String(data: data.prefix(1200), encoding: .utf8) ?? "<non-utf8 body>"
        NSLog("GeoSphere body preview: %@", bodyPreview)
    }
}

private struct DailyHeatWarning {
    let level: Int
    let start: Date?
    let end: Date?
}

private struct GeoSphereResolvedState {
    let municipalityID: String
    let municipalityName: String
    let warnings: [GeoSphereWarning]
}

private struct GeoSphereLookupResponse: Decodable {
    let type: String
    let msg: String?
    let properties: GeoSphereLookupProperties?
}

private struct GeoSphereLookupProperties: Decodable {
    let location: GeoSphereLocation?
    let warnings: [GeoSphereWarning]?
}

private struct GeoSphereLocation: Decodable {
    let properties: GeoSphereMunicipality?
}

private struct GeoSphereMunicipality: Decodable {
    let gemeindenr: GeoSphereIDValue
    let name: String?
}

private struct GeoSphereIDValue: Decodable {
    let value: String?

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()

        if let intValue = try? container.decode(Int.self) {
            value = String(intValue)
            return
        }

        if let stringValue = try? container.decode(String.self) {
            value = stringValue
            return
        }

        value = nil
    }
}

private struct GeoSphereWarning: Decodable {
    let properties: GeoSphereWarningProperties
}

private struct GeoSphereWarningProperties: Decodable {
    let warntypid: Int?
    let warnstufeid: Int?
    let wtype: Int?
    let wlevel: Int?
    let rawinfo: GeoSphereRawInfo?
    let rawfinfo: GeoSphereRawInfo?
    let warningType: String?
    let begin: String?
    let start: String?
    let end: String?

    enum CodingKeys: String, CodingKey {
        case warntypid
        case warnstufeid
        case wtype
        case wlevel
        case rawinfo
        case rawfinfo
        case warningType = "warning_type"
        case begin
        case start
        case end
    }
}

private struct GeoSphereRawInfo: Decodable {
    let wtype: Int?
    let wlevel: Int?
    let start: String?
    let end: String?
}

private struct OpenMeteoForecastResponse: Decodable {
    let daily: OpenMeteoDaily
}

private struct OpenMeteoDaily: Decodable {
    let time: [String]
    let uvIndexMax: [Double?]
    let apparentTemperatureMax: [Double?]

    enum CodingKeys: String, CodingKey {
        case time
        case uvIndexMax = "uv_index_max"
        case apparentTemperatureMax = "apparent_temperature_max"
    }
}

File: Android/app/src/main/java/org/entner/HitzeV/data/DashboardDataService.kt
package org.entner.HitzeV.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.entner.HitzeV.model.DailyForecast
import org.entner.HitzeV.model.GeoCoordinate
import org.entner.HitzeV.model.HazardSeverity
import org.entner.HitzeV.model.WarningTimeRange
import org.entner.HitzeV.model.WorksiteSnapshot
import java.net.HttpURLConnection
import java.net.URL
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.util.Locale

class DashboardDataService(
    private val appStorage: AppStorage
) {
    private val json = Json { ignoreUnknownKeys = true }
    private val defaultGeosphereBaseUrl = "https://warnungen.zamg.at/wsapp/api/getWarningsForCoords"
    private val openMeteoBaseUrl = "https://api.open-meteo.com/v1/forecast"
    private val viennaZoneId = ZoneId.of("Europe/Vienna")

    suspend fun fetchSnapshot(coordinate: GeoCoordinate): WorksiteSnapshot = coroutineScope {
        val geoSphereResult = async { fetchGeoSphere(coordinate) }
        val openMeteoResult = async { fetchOpenMeteo(coordinate) }

        val geoResult = geoSphereResult.await()
        val meteoResult = openMeteoResult.await()

        val today = LocalDate.now(viennaZoneId)
        val forecasts = (0..3).map { offset ->
            val date = today.plusDays(offset.toLong())
            val severity = highestHazardSeverity(date, geoResult.warnings)
            val warningTimeRanges = warningTimeRangesForDate(date, geoResult.warnings)
            val uvIndex = meteoResult.daily.uvIndexMax.getOrNull(offset)
            val apparentTemperature = meteoResult.daily.apparentTemperatureMax.getOrNull(offset)

            DailyForecast(
                id = date.toString(),
                date = date,
                severity = severity,
                apparentTemperatureMax = apparentTemperature,
                uvIndexMax = uvIndex,
                warningTimeRanges = warningTimeRanges
            )
        }

        val todayForecast = forecasts.firstOrNull()
        WorksiteSnapshot(
            municipalityId = geoResult.municipalityId,
            municipalityName = geoResult.municipalityName,
            severity = todayForecast?.severity ?: HazardSeverity.NONE,
            uvIndex = todayForecast?.uvIndexMax,
            apparentTemperature = todayForecast?.apparentTemperatureMax,
            forecasts = forecasts,
            updatedAt = Instant.now()
        )
    }

    suspend fun resolveMunicipalityForCoordinate(coordinate: GeoCoordinate): ResolvedMunicipality {
        return fetchGeoSphere(coordinate)
    }

    private suspend fun fetchGeoSphere(coordinate: GeoCoordinate): ResolvedMunicipality {
        val geosphereBaseUrl = appStorage.customGeoSphereUrl.first().ifBlank { defaultGeosphereBaseUrl }
        val url = "$geosphereBaseUrl?lat=${coordinate.latitude.asQueryValue()}&lon=${coordinate.longitude.asQueryValue()}"
        val root = performGet(url)

        if (root["type"].asStringOrNull()?.equals("error", ignoreCase = true) == true) {
            throw DashboardDataError.MunicipalityNotFound(
                root["msg"].asStringOrNull() ?: "Could not find municipality for coordinates."
            )
        }

        val properties = root["properties"]?.jsonObject ?: throw DashboardDataError.InvalidGeoSphereResponse
        val municipalityProperties = properties["location"]?.jsonObject
            ?.get("properties")
            ?.jsonObject ?: throw DashboardDataError.InvalidGeoSphereResponse

        val municipalityId = municipalityProperties["gemeindenr"].municipalityValue()
            ?: throw DashboardDataError.InvalidGeoSphereResponse
        val municipalityName = municipalityProperties["name"].asStringOrNull() ?: "Gemeinde $municipalityId"

        val warnings = properties["warnings"]?.jsonArray?.mapNotNull { warningElement ->
            parseWarning(warningElement.jsonObject)
        }.orEmpty()

        return ResolvedMunicipality(
            municipalityId = municipalityId,
            municipalityName = municipalityName,
            warnings = warnings
        )
    }

    private suspend fun fetchOpenMeteo(coordinate: GeoCoordinate): OpenMeteoForecastResponse {
        val url = buildString {
            append(openMeteoBaseUrl)
            append("?latitude=${coordinate.latitude.asQueryValue()}")
            append("&longitude=${coordinate.longitude.asQueryValue()}")
            append("&daily=uv_index_max,apparent_temperature_max")
            append("&timezone=Europe/Vienna")
            append("&forecast_days=4")
        }

        val root = performGet(url)
        return json.decodeFromString(root.toString())
    }

    private suspend fun performGet(urlString: String): JsonObject = withContext(Dispatchers.IO) {
        val connection = (URL(urlString).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 10_000
            readTimeout = 10_000
            doInput = true
        }

        try {
            val statusCode = connection.responseCode
            if (statusCode !in 200..299) {
                val errorBody = connection.errorStream?.bufferedReader()?.use { it.readText() }.orEmpty()
                val parsedMunicipalityMessage = runCatching {
                    json.parseToJsonElement(errorBody).jsonObject
                }.getOrNull()?.municipalityNotFoundMessage()
                if (!parsedMunicipalityMessage.isNullOrBlank()) {
                    throw DashboardDataError.MunicipalityNotFound(parsedMunicipalityMessage)
                }
                throw DashboardDataError.Network("HTTP $statusCode")
            }

            val body = connection.inputStream.bufferedReader().use { it.readText() }
            json.parseToJsonElement(body).jsonObject
        } catch (error: DashboardDataError) {
            throw error
        } catch (error: Exception) {
            throw DashboardDataError.Network(error.localizedMessage ?: "Unknown network error")
        } finally {
            connection.disconnect()
        }
    }

    private fun parseWarning(objectValue: JsonObject): GeoSphereWarning? {
        val properties = objectValue["properties"]?.jsonObject ?: return null
        val warningTypeCandidates = listOfNotNull(
            properties["wtype"].asInt(),
            properties["warntypid"].asInt(),
            properties["rawinfo"]?.jsonObject?.get("wtype").asInt(),
            properties["rawfinfo"]?.jsonObject?.get("wtype").asInt()
        )
        val warningLevelCandidates = listOfNotNull(
            properties["warnstufeid"].asInt(),
            properties["wlevel"].asInt(),
            properties["rawinfo"]?.jsonObject?.get("wlevel").asInt(),
            properties["rawfinfo"]?.jsonObject?.get("wlevel").asInt()
        )

        return GeoSphereWarning(
            warningTypeId = warningTypeCandidates.firstOrNull(),
            warningTypeCandidates = warningTypeCandidates,
            warningLevel = warningLevelCandidates.maxOrNull() ?: 0,
            warningTypeText = properties["warning_type"].asStringOrNull(),
            start = properties["start"].asStringOrNull()
                ?: properties["begin"].asStringOrNull()
                ?: properties["rawinfo"]?.jsonObject?.get("start").asStringOrNull()
                ?: properties["rawfinfo"]?.jsonObject?.get("start").asStringOrNull(),
            end = properties["end"].asStringOrNull()
                ?: properties["rawinfo"]?.jsonObject?.get("end").asStringOrNull()
                ?: properties["rawfinfo"]?.jsonObject?.get("end").asStringOrNull()
        )
    }

    data class ResolvedMunicipality(
        val municipalityId: String,
        val municipalityName: String,
        val warnings: List<GeoSphereWarning>
    )

    data class GeoSphereWarning(
        val warningTypeId: Int?,
        val warningTypeCandidates: List<Int>,
        val warningLevel: Int,
        val warningTypeText: String?,
        val start: String?,
        val end: String?
    )

    @Serializable
    private data class OpenMeteoForecastResponse(
        val daily: OpenMeteoDaily = OpenMeteoDaily()
    )

    @Serializable
    private data class OpenMeteoDaily(
        val time: List<String> = emptyList(),
        @SerialName("uv_index_max")
        val uvIndexMax: List<Double?> = emptyList(),
        @SerialName("apparent_temperature_max")
        val apparentTemperatureMax: List<Double?> = emptyList()
    )

    companion object {
        private val localGeosphereDateTimeFormatters = listOf(
            DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm", Locale.GERMANY),
            DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm", Locale.US)
        )

        private data class DailyHeatWarning(
            val level: Int,
            val start: Instant?,
            val end: Instant?
        )

        internal fun parseGeosphereTimestamp(value: String?): Instant? {
            val trimmed = value?.trim().orEmpty()
            if (trimmed.isEmpty()) return null

            trimmed.toDoubleOrNull()?.let { numericValue ->
                val epochMillis = if (numericValue > 100_000_000_000) {
                    numericValue.toLong()
                } else {
                    (numericValue * 1000).toLong()
                }
                return Instant.ofEpochMilli(epochMillis)
            }

            return try {
                Instant.parse(trimmed)
            } catch (_: DateTimeParseException) {
                localGeosphereDateTimeFormatters.firstNotNullOfOrNull { formatter ->
                    runCatching {
                        LocalDateTime.parse(trimmed, formatter)
                            .atZone(ZoneId.of("Europe/Vienna"))
                            .toInstant()
                    }.getOrNull()
                }
            }
        }

        internal fun highestHazardSeverity(
            targetDate: LocalDate,
            warnings: List<GeoSphereWarning>,
            zoneId: ZoneId = ZoneId.of("Europe/Vienna")
        ): HazardSeverity {
            val highestHeat = heatWarningsForDate(targetDate, warnings, zoneId)
                .maxOfOrNull(DailyHeatWarning::level) ?: 0
            return if (highestHeat > 0) HazardSeverity.heat(highestHeat) else HazardSeverity.NONE
        }

        internal fun warningTimeRangesForDate(
            targetDate: LocalDate,
            warnings: List<GeoSphereWarning>,
            zoneId: ZoneId = ZoneId.of("Europe/Vienna")
        ): List<WarningTimeRange> {
            val startOfDay = targetDate.atStartOfDay(zoneId).toInstant()
            val endOfDay = targetDate.plusDays(1).atStartOfDay(zoneId).minusNanos(1).toInstant()

            return heatWarningsForDate(targetDate, warnings, zoneId)
                .map { warning ->
                    WarningTimeRange(
                        start = maxOf(warning.start ?: startOfDay, startOfDay),
                        end = minOf(warning.end ?: endOfDay, endOfDay)
                    )
                }
                .sortedBy(WarningTimeRange::start)
                .fold(mutableListOf()) { merged, range ->
                    val lastRange = merged.lastOrNull()
                    if (lastRange == null || range.start.isAfter(lastRange.end)) {
                        merged += range
                    } else {
                        merged[merged.lastIndex] = lastRange.copy(end = maxOf(lastRange.end, range.end))
                    }
                    merged
                }
        }

        private fun heatWarningsForDate(
            targetDate: LocalDate,
            warnings: List<GeoSphereWarning>,
            zoneId: ZoneId
        ): List<DailyHeatWarning> {
            val startOfDay = targetDate.atStartOfDay(zoneId).toInstant()
            val endOfDay = targetDate.plusDays(1).atStartOfDay(zoneId).minusNanos(1).toInstant()

            return warnings.mapNotNull { warning ->
                val startDate = parseGeosphereTimestamp(warning.start)
                val endDate = parseGeosphereTimestamp(warning.end)
                val intersects = when {
                    startDate != null && endDate != null -> !(endDate.isBefore(startOfDay) || startDate.isAfter(endOfDay))
                    startDate != null -> !startDate.isAfter(endOfDay)
                    endDate != null -> !endDate.isBefore(startOfDay)
                    else -> true
                }

                if (!intersects || !isHeatWarning(warning)) {
                    null
                } else {
                    DailyHeatWarning(
                        level = warning.warningLevel,
                        start = startDate,
                        end = endDate
                    )
                }
            }
        }

        private fun isHeatWarning(warning: GeoSphereWarning): Boolean {
            if (warning.warningTypeCandidates.contains(6)) return true

            val type = warning.warningTypeId
            if (type != null) return false

            val textualType = warning.warningTypeText?.lowercase(Locale.ROOT).orEmpty()
            return textualType.contains("hitze") || textualType.contains("heat")
        }
    }
}

sealed class DashboardDataError(message: String) : Exception(message) {
    data object InvalidRequest : DashboardDataError("Ungueltige Anfrageparameter.")
    data object InvalidGeoSphereResponse : DashboardDataError("GeoSphere Antwort konnte nicht gelesen werden.")
    data class MunicipalityNotFound(val detail: String) : DashboardDataError("Gemeinde nicht gefunden: $detail")
    data class Network(val detail: String) : DashboardDataError("Netzwerkfehler: $detail")
}

private fun Double.asQueryValue(): String = String.format(Locale.US, "%.6f", this)

private fun JsonElement?.asInt(): Int? = this?.jsonPrimitive?.intOrNull

private fun JsonElement?.municipalityValue(): String? {
    val element = this ?: return null
    return if (element is JsonObject) element["value"].municipalityValue() else element.asStringOrNull()
}

private fun JsonElement?.asStringOrNull(): String? =
    runCatching { this?.jsonPrimitive?.content }
        .getOrNull()
        ?.takeIf { it.isNotBlank() }

private fun JsonObject.municipalityNotFoundMessage(): String? {
    val type = this["type"].asStringOrNull()?.lowercase(Locale.ROOT)
    if (type != "error") return null

    val message = this["msg"].asStringOrNull() ?: return "Could not find municipal for coords."
    val normalizedMessage = message.lowercase(Locale.ROOT)
    return if (
        normalizedMessage.contains("could not find municipal for coords") ||
        normalizedMessage.contains("could not find municipality") ||
        normalizedMessage.contains("municipal for coords")
    ) {
        message
    } else {
        null
    }
}
